File: 2021/11/27/max_product.py
"""
https://leetcode.com/problems/shuffle-the-array/
Author: Andrew Jarombek
Date: 11/27/2021
"""

from typing import List


def max_product(nums: List[int]) -> int:
    first = 0
    second = 0

    for num in nums:
        if num > first:
            second = first
            first = num
        elif num > second:
            second = num

    return (first - 1) * (second - 1)

File: 2022/12/15/maxProduct.js
/**
 * https://leetcode.com/problems/maximum-product-of-three-numbers/
 * @author Andrew Jarombek
 * @since 12/15/2022
 */

const maximumProduct = function(nums) {
    nums.sort((x, y) => x - y);

    if (nums[nums.length - 1] < 0) {
        return nums.slice(nums.length - 3).reduce((acc, val) => acc * val, 1);
    }

    return Math.max(
        nums[0] * nums[1] * nums[nums.length - 1],
        nums[nums.length - 3] * nums[nums.length - 2] * nums[nums.length - 1]
    );
};

console.assert(maximumProduct([1, 2, 3, 4]) === 24);
console.assert(maximumProduct([-100, -98, -1, 2, 3, 4]) === 39200);
